File: python/day14_part2_2020.py

import re
from itertools import product

def apply_mask_v1(value, mask):
    value_bin = format(value, '036b')
    result = [value_bin[i] if mask[i] == 'X' else mask[i] for i in range(36)]
    return int(''.join(result), 2)

def apply_mask_v2(address, mask):
    address_bin = format(address, '036b')
    result = ['1' if mask[i] == '1' else address_bin[i] if mask[i] == '0' else 'X' for i in range(36)]
    floating_indices = [i for i, bit in enumerate(result) if bit == 'X']
    floating_combinations = product(['0', '1'], repeat=len(floating_indices))
    addresses = []
    for combination in floating_combinations:
        temp_address = result.copy()
        for i, c in zip(floating_indices, combination):
            temp_address[i] = c
        addresses.append(int(''.join(temp_address), 2))
    return addresses

def main():
    memory = {}
    mask = ''
    with open('input.txt', 'r') as file:
        for line in file:
            if line.startswith('mask'):
                mask = line.split(' = ')[1].strip()
            else:
                address, value = map(int, re.findall(r'\d+', line))
                addresses = apply_mask_v2(address, mask)
                for addr in addresses:
                    memory[addr] = value

    print(sum(memory.values()))

if __name__ == "__main__":
    main()

File: training_data/reflection-tuning/day14_part2_2020.rb
def apply_mask_v1(value, mask)
  mask.chars.reverse.each_with_index do |bit, i|
    case bit
    when '0' then value &= ~(1 << i)
    when '1' then value |= (1 << i)
    end
  end
  value
end

def apply_mask_v2(address, mask, value, memory, index = 0)
  if index == 36
    memory[address] = value
    return
  end

  case mask[index]
  when '0'
    apply_mask_v2(address, mask, value, memory, index + 1)
  when '1'
    apply_mask_v2(address | (1 << (35 - index)), mask, value, memory, index + 1)
  when 'X'
    apply_mask_v2(address & ~(1 << (35 - index)), mask, value, memory, index + 1)
    apply_mask_v2(address | (1 << (35 - index)), mask, value, memory, index + 1)
  end
end

def solve(input)
  memory_v1 = {}
  memory_v2 = {}
  mask = ''

  input.each_line do |line|
    if line.start_with?('mask')
      mask = line.split(' = ').last.strip
    else
      address, value = line.scan(/\d+/).map(&:to_i)
      memory_v1[address] = apply_mask_v1(value, mask)
      apply_mask_v2(address, mask, value, memory_v2)
    end
  end

  [memory_v1.values.sum, memory_v2.values.sum]
end

input = File.read('input.txt')
part1, part2 = solve(input)
puts "Part 1: #{part1}"
puts "Part 2: #{part2}"
